libp2p/network/connection/swarm_connection.py

```diff
@@ -0,0 +1,76 @@
+import asyncio
+from typing import TYPE_CHECKING, Any, Awaitable, List, Set, Tuple
+
+from libp2p.network.connection.net_connection_interface import INetConn
+from libp2p.network.stream.net_stream import NetStream
+from libp2p.stream_muxer.abc import IMuxedConn, IMuxedStream
+
+if TYPE_CHECKING:
+    from libp2p.network.swarm import Swarm  # noqa: F401
+
+
+"""
+Reference: https://github.com/libp2p/go-libp2p-swarm/blob/04c86bbdafd390651cb2ee14e334f7caeedad722/swarm_conn.go  # noqa: E501
+"""
+
+
+class SwarmConn(INetConn):
+    conn: IMuxedConn
+    swarm: "Swarm"
+    streams: Set[NetStream]
+    event_closed: asyncio.Event
+
+    _tasks: List["asyncio.Future[Any]"]
+
+    def __init__(self, conn: IMuxedConn, swarm: "Swarm") -> None:
+        self.conn = conn
+        self.swarm = swarm
+        self.streams = set()
+        self.event_closed = asyncio.Event()
+
+        self._tasks = []
+
+    async def close(self) -> None:
+        if self.event_closed.is_set():
+            return
+        self.event_closed.set()
+        await self.conn.close()
+        for task in self._tasks:
+            task.cancel()
+
+        # TODO: Reset streams for local.
+        # TODO: Notify closed.
+
+    async def _handle_new_streams(self) -> None:
+        # TODO: Break the loop when anything wrong in the connection.
+        while True:
```

Comment:
does it make sense to put this `while` block inside a `try` block and then put `await self.close()` inside the `finally:` clause?

Reply:
It seems pretty convenient but is it also possible that we catch unwanted exceptions? In the latest code I catch only the `MuxedConnUnavailable` in this loop.

https://github.com/libp2p/py-libp2p/pull/304/files#diff-cc33a5151ea2f5acfde91e1dd54be5d4R52-R63

Reply:
On second thought, I'm thinking it might be possibly an easier pattern for the shutting down logic in `Mplex` in #304 , will investigate more. Thanks for bringing this up.